File: frontend/src/schemas.ts
// The API schemas

export type Project = {
  id: number;
  name: string;
  description_1: string;
  description_2: string;
  fixed: boolean;
  min_points: number;
  max_points: number;
  rank: number;
  points: number;
  points_text: string;
  marked: boolean;
};

export type VotedProject = {
  id: number;
  rank: number;
  points: number;
  marked: boolean;
};

export type DataResponse = {
  voted: boolean;
  max_total_points: number;
  points_step: number;
  projects: Project[];
};

export type VoteRequestBody = {
  projects: VotedProject[];
};

File: backend/src/schemas.py
# contains the API schemas
# schema - object that the API return or receive

from pydantic import BaseModel, Field


class ProjectSchema(BaseModel):
    """Project with the data of voter."""

    id: int
    name: str
    description_1: str
    description_2: str
    fixed: bool
    min_points: int
    max_points: int
    rank: int
    points: int
    points_text: str
    marked: bool


class VotedProjectSchema(BaseModel):
    id: int
    rank: int
    points: int = Field(ge=0)  # ensures points >= 0
    marked: bool


class DataResponseSchema(BaseModel):
    voted: bool
    max_total_points: int
    points_step: int
    open_for_voting: bool
    results: dict | None
    note: str
    projects: list[ProjectSchema]


class VoteRequestBodySchema(BaseModel):
    note: str
    projects: list[VotedProjectSchema]
